fdbserver/GetCipherKeys.actor.cpp

```diff
@@ -0,0 +1,248 @@
+	state std::unordered_map<BlobCipherDetails, Reference<BlobCipherKey>> cipherKeys;
+	state std::unordered_set<BaseCipherIndex, boost::hash<BaseCipherIndex>> uncachedBaseCipherIds;
+	state EKPGetBaseCipherKeysByIdsRequest request;
+
+	if (!db.isValid()) {
+		TraceEvent(SevError, "GetCipherKeys_ServerDBInfoNotAvailable");
+		throw encrypt_ops_error();
+	}
+
+	// Collect cached cipher keys.
+	for (const BlobCipherDetails& details : cipherDetails) {
+		Reference<BlobCipherKey> cachedCipherKey =
+		    cipherKeyCache->getCipherKey(details.encryptDomainId, details.baseCipherId, details.salt);
+		if (cachedCipherKey.isValid()) {
+			cipherKeys.emplace(details, cachedCipherKey);
+		} else {
+			uncachedBaseCipherIds.insert(std::make_pair(details.encryptDomainId, details.baseCipherId));
+		}
+	}
+	if (uncachedBaseCipherIds.empty()) {
+		return cipherKeys;
+	}
+	for (const BaseCipherIndex& id : uncachedBaseCipherIds) {
+		request.baseCipherInfos.emplace_back(
+		    id.first /*domainId*/, id.second /*baseCipherId*/, StringRef() /*domainName*/, request.arena);
+	}
+	// Fetch any uncached cipher keys.
+	loop choose {
+		when(std::vector<EKPBaseCipherDetails> baseCipherDetails = wait(getUncachedCipherKeys(db, request))) {
+			std::unordered_map<BaseCipherIndex, StringRef, boost::hash<BaseCipherIndex>> baseCipherKeys;
+			for (const EKPBaseCipherDetails& baseDetails : baseCipherDetails) {
+				BaseCipherIndex baseIdx = std::make_pair(baseDetails.encryptDomainId, baseDetails.baseCipherId);
+				baseCipherKeys[baseIdx] = baseDetails.baseCipherKey;
+			}
+			// Insert base cipher keys into cache and construct result.
+			for (const BlobCipherDetails& details : cipherDetails) {
+				if (cipherKeys.count(details) > 0) {
+					continue;
+				}
+				BaseCipherIndex baseIdx = std::make_pair(details.encryptDomainId, details.baseCipherId);
+				const auto& itr = baseCipherKeys.find(baseIdx);
+				if (itr == baseCipherKeys.end()) {
+					TraceEvent(SevError, "GetCipherKeys_KeyMissing")
+					    .detail("DomainId", details.encryptDomainId)
+					    .detail("BaseCipherId", details.baseCipherId);
+					throw encrypt_key_not_found();
+				}
+				Reference<BlobCipherKey> cipherKey = cipherKeyCache->insertCipherKey(details.encryptDomainId,
+				                                                                     details.baseCipherId,
+				                                                                     itr->second.begin(),
+				                                                                     itr->second.size(),
+				                                                                     details.salt);
+				ASSERT(cipherKey.isValid());
+				cipherKeys[details] = cipherKey;
+			}
+			break;
+		}
+		// In case encryptKeyProxy has changed, retry the request.
+		when(wait(onEncryptKeyProxyChange(db))) {}
+	}
+	return cipherKeys;
+}
+
+ACTOR Future<TextAndHeaderCipherKeys> getLatestSystemCipherKeys(Reference<AsyncVar<ServerDBInfo> const> db) {
```

Comment:
a nit: `getLatestSystemCipherKeys` doesn't relate too well with the routine; in essence the function fetches `defaultDomainCipherKeys` (not specific to any tenant). 

But, yes I don't have any good suggestion; let me think more.

Reply:
The corresponding domain id is named `SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID` and the function name here is consistent with that. Do you prefer changing both to "default_xxx" instead of "system_xxx"?